src/storage/txn/latch.rs

```diff
@@ -123,25 +182,34 @@ impl Latches {
     /// Preconditions: the caller must ensure the command is at the front of the latches.
     pub fn release(&self, lock: &Lock, who: u64) -> Vec<u64> {
         let mut wakeup_list: Vec<u64> = vec![];
-        for i in &lock.required_slots[..lock.owned_count] {
-            let mut latch = self.slots[*i].lock().unwrap();
-            let front = latch.waiting.pop_front().unwrap();
+        for &key_hash in &lock.required_hashes[..lock.owned_count] {
+            let mut latch = self.lock_latch(key_hash);
+            let (v, front) = latch.pop_front(key_hash).unwrap();
             assert_eq!(front, who);
-            if let Some(wakeup) = latch.waiting.front() {
-                wakeup_list.push(*wakeup);
+            assert_eq!(v, key_hash);
+            if let Some(wakeup) = latch.get_first_req_by_hash(key_hash) {
+                wakeup_list.push(wakeup);
             }
         }
         wakeup_list
     }
 
-    /// Calculates the slot ID by hashing the `key`.
-    fn calc_slot<H>(&self, key: &H) -> usize
+    /// Calculates the hash value of the `key`.
+    fn calc_slot<H>(&self, key: &H) -> u64
     where
         H: Hash,
     {
         let mut s = DefaultHasher::new();
         key.hash(&mut s);
-        (s.finish() as usize) & (self.size - 1)
+
+        s.finish()
+    }
+
+    #[inline]
+    fn lock_latch(&self, hash: u64) -> MutexGuard<'_, Latch> {
+        self.slots[(hash as usize) & (self.size - 1)]
```

Comment:
```suggestion
        self.slots[(hash as usize) % self.size]
```

What about remainder to evenly distribute hashes?